### PR TITLE
feat(perf): Improve N+1 API Calls detector

### PR DESCRIPTION
**Pull request by gggritso**
The N+1 API Calls detector has no similarity criterion, which is a problem. It flags _any_ set of more than five concurrent `http.client` spans. _Lots_ of our transactions make >5 concurrent requests for good reasons, though, if they need a lot of data! For example, `/organizations/:orgId/issues/:groupId/`. I originally shipped this detector with no similarity criterion because I was hoping that the requirement of 5 concurrent requests is strict enough, but it is not.

We need _some_ kind of way to detect if the spans are _similar_, which they would be in a true N+1 error. Span hashes to the rescue. [Span hashes for `http.client`](https://github.com/getsentry/sentry/pull/28184/files#diff-c11ec6bbfcb4c66f3b913b786ad2ceb78cc5ce73dbd7fc25f103fb099fb6776eR146-R181) spans are an md5 digest of the URL without the URL parameters. That's a great start for us! This should be enough to catch multiple concurrent requests to the same endpoint. In the future we can write our own similarity checking (then we can catch interpolated URLs like `/resources/773428/data`).

## Changes

- Tighten up the requirement for similarity. Spans must be concurrent _and_ have the same hash, which is equivalent to requiring them to have the same URL path
- Loosen up the requirement for concurrency. 3 concurrent spans or more to the same URL will be an issue, instead of 5. Now that we're checking similarity, we can loosen this up
- Remove the `http.server` op type. `http.server` doesn't use the same hashing strategy, so I'm leaving it out

Anecdotally, this approach will flag transactions like `/organizations/:orgId/dashboard/:dashboardId/` where multiple widgets are requesting event data and event stats for different data sets, but the same time range and projects.

### CODE REVIEW

File: src/sentry/utils/performance_issues/performance_detection.py [hunk 1]
```diff
@@ -525,25 +526,31 @@ def visit_span(self, span: Span):
         if not fingerprint:
             return
 
+        hash = span.get("hash", None)
+        if not hash:
+            return
```
Comment (k-fish) [inline src/sentry/utils/performance_issues/performance_detection.py:531]: Have you considered replacing the broad `fingerprint` on the op w/ just the hash? Then we can do  
```py
most_recent_start =  self.most_recent_start_time[hash]
delta_to_previous_span_start_time =  span["start_timestamp"] - most_recent_start
```
Then each span would look up it's last seen similar span time.

**Review by k-fish — APPROVED** [review]
👍, not sure if you want to address my comment or just try it as is.

File: src/sentry/utils/performance_issues/performance_detection.py (hunk 1 above)
Comment (gggritso) [reply src/sentry/utils/performance_issues/performance_detection.py:531]: @k-fish N+1 API calls only applies to one span op anyway, so I don't need fingerprinting at all. I could just use a single array, and single values! I opted to keep the code similar to the other detectors, for readability, so I did the fingerprinting thing. What do you think, is the hash thing worth it, or is it better to keep the detectors looking similar for now?

Comment (k-fish) [reply src/sentry/utils/performance_issues/performance_detection.py:531]: We definitely should be using the `span_group` hash for all of our detectors moving forward and just improve that logic to help both perf issues and suspect spans, if we have good ideas for better detecting 'similarity'

Comment (gggritso) [reply src/sentry/utils/performance_issues/performance_detection.py:531]: Makes sense!

**Pull request merged**